### ISSUE TITLE
"Extending Embedded Python" Documentation/Safe API

### ISSUE THREAD

**Issue by Infrasonics** [PyO3/pyo3#1983]
# Request
Add documentation for the extension of an embedded Python interpreter to the guide. Maybe it makes sense to also provide a safe API e.g. for `PyImport_AppendInittab` and family.

# Use Case
I tried to extend the interpreter with an internal module so that externally loaded Python scripts would be able to use that module to interact with the application they are running in. Since this module is only meaningful while the scripts are running in the application's context, it does not make sense to build the extension module as e.g. a DLL.

For this use case the Python C-API has [this example](https://docs.python.org/3/extending/embedding.html#extending-embedded-python). I got quickly stuck in my actual case and decided to rebuild exactly this example (my current solution can be found [here](https://github.com/Infrasonics/pyo3-AppendInittab-example).

## Context
``` rust
#[pymodule]
fn emb(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(numargs, m)?)?;
    Ok(())
}
```
[Listing 1: Module definition with macro]
Where the `numargs` function is regularly defined with `#[pyfunction]`.

# Possible Solutions
## Exploiting the Behavior of `#[pymodule]`
This distilles down to the following:
1. Make use of the fact that `#[pymodule]` creates a corresponding `PyInit_<funcname>` function in a proc_macro, which can then be used with
``` rust
unsafe {
    let ptr = std::mem::transmute::<*const (), extern "C" fn() -> *mut ffi::PyObject>(
        PyInit_emb as *const ()
    );
    ffi::PyImport_AppendInittab("emb\0".as_ptr() as *const _, Some(ptr));
}
```
[Listing 2: Interpreter extension pattern]

2. A safe version of `PyImport_AppendInittab` would be nice for that. Or maybe an additional attribute to extend the current interpreter with that module transparently?
3. This only works for single phase initialization since it uses `PyModule_Create` internally.
4. If this is an acceptable solution, I'd love to find that in the guide/docs next time I'm looking for it.

### Advantages
- Easy to use, only needs documentation
- Little unsafe code
- Little boilerplate

### Disadvantages
- Depends on an implementation detail
- Only works for single phase initialization

## Manually Creating the `PyInit_` Function
Going the manual way, not using `#[pymodule]`. What I'm showing next is basically what the `#[pymodule]` macro generates with the exclusion of the `RefCell` wrapper for `PyModuleDef` from the derive_utils. From what I tried and read, this looks roughly like:
``` rust
#[no_mangle]
#[allow(non_snake_case)]
pub unsafe extern "C" fn PyInit_emb() -> *mut ::pyo3::ffi::PyObject {
        let fptr = std::mem::transmute::<*const (), ffi::PyCFunction>(numargs as *const ());
        let method_numargs = ffi::PyMethodDef {
            ml_name: cstr!("numargs\0"),
            ml_meth: Some(fptr),
            ml_flags: ffi::METH_VARARGS,
            ml_doc: cstr!("Return the number of arguments received by the process\0"),
        };

        let method_sentinel = ffi::PyMethodDef {
            ml_name: std::ptr::null_mut(),
            ml_meth: None,
            ml_flags: 0,
            ml_doc: std::ptr::null_mut(),
        };

        let mut methods_array: [ffi::PyMethodDef; 2] = [method_numargs, method_sentinel];
        let methods: *mut ffi::PyMethodDef = methods_array.as_mut_ptr();

        let def: *mut ffi::PyModuleDef = ffi::PyModuleDef {
            m_base: ffi::PyModuleDef_HEAD_INIT,
            m_name: cstr!("emb\0"),
            m_doc: cstr!("somedoc\0"),
            m_size: 0,
            m_methods: methods,
            m_slots: std::ptr::null_mut(),
            m_traverse: None,
            m_clear: None,
            m_free: None,
        }
        .borrow_mut();

        ::pyo3::callback::handle_panic(|_py| {
            let module = unsafe { _py.from_owned_ptr_or_err::<PyModule>(ffi::PyModule_Create(def))? };
            module.add_function(wrap_pyfunction!(numargs, module)?)?;
            Ok(pyo3::IntoPyPointer::into_ptr(module))
        })
}
```
[Listing 3: Manual ffi-based module definition]

Using this instead of the code shown in Listing 1 results for me in this runtime error:
`Error: SystemError: PyState_AddModule called on module with slots`.
1. I still don't understand why this happens even though `m_slots` is effectively `NULL` and where the difference to the expanded macros is, so if anyone could explain this to me, I'd be grateful.
2. This requires a lot of unsafe code and rather unergonomic pointer casting/handling, so some design to enable this in a smoother way would be nice to have here.
3. With some macro magic, I'm sure it could be sufficiently facilitated.

### Advantages
- Independant of macro implementation details
- More versatile, e.g. also for multi-phase initialization and therefore
  multi-interpreter possible

### Disadvantages
- basically using `ffi` only
- from my point of view, a lot of documentation to find and read to use it
- not ergonomic at all (lots of unsafe code and boilerplate)

## Next Alternative
Since I'm very new to PyO3, there most probably is a much better way of doing this. I was not able to find it though and therefore request a section in the guide/docs on how to get that behavior.

Overall PyO3 was a pleasure to use for me so far and this was the only major headache until now.
Thanks for this library and keep up the good work.

**Comment by davidhewitt** [comment]
We do now have `append_to_inittab` macro, so I'm going to close this issue as it's quite old now.
